File: ros/ws01/first_ws/src/service_pkg/src/server_custom_srv3.cpp
//
// Created by zcb on 2021/4/9.
//


#include <ros/ros.h>
#include <iostream>
#include <demo_srvs/FindTeam.h>

using namespace std;

bool client_callback(demo_srvs::FindTeamRequest &request, demo_srvs::FindTeamResponse &response) {
    string name = request.stu.name;
    int age = request.stu.age;

    response.team.name = "NCHU";
    response.team.leader.name = name;
    response.team.leader.age = age;

    return true;
}


int main(int argc, char **argv) {

    string nodeName = "cpp_server_node";
    ros::init(argc, argv, nodeName);
    ros::NodeHandle node;

    string serviceName = "/cpp/service";
    const ros::ServiceServer &server = node.advertiseService(serviceName, client_callback);

    ros::spin();

    return 0;
}

File: 06ros/ws01/first_ws/src/service_pkg/scripts/server_custom_srv3.py
#! /usr/bin/env python
# coding:utf8

import rospy
from demo_srvs.srv import FindTeam, FindTeamRequest, FindTeamResponse


def client_callback(request):
    if not isinstance(request, FindTeamRequest):
        return

    response = FindTeamResponse()
    name = request.stu.name
    age = request.stu.age
    response.team.name = "NCHU"
    response.team.leader.name = name
    response.team.leader.age = age

    return response


if __name__ == '__main__':
    nodeName = "python_service_node"
    rospy.init_node(nodeName)

    serviceName = "/python/service"

    # 与c++ api 不一致，
    server = rospy.Service(serviceName, FindTeam, client_callback)

    rospy.spin()
